### PR TITLE
vmm: Allow simultaneously set serial and console as TTY mode

### PR DESCRIPTION
**Pull request by zhuangel**
Performance of legacy serial is poor, so we want to use virtio console for better performance, but once we configure VM console as virito console, the boot stage kernel logs will be lost, including some critical exception message in boot stage.

Linux support assign console for early stage via "earlyprintk", with this feature we could assign legacy serial to early stage, and virito console to normal console (after early boot stage), then we could catch most VM console logs via virito console, also catch boot stage important logs via legacy serial.

This patch remove the double TTY restriction, so that we could use both serial and console with TTY mode.

### CODE REVIEW

File: vmm/src/config.rs [hunk 1]
```diff
@@ -1930,9 +1930,18 @@ impl VmConfig {
...
+        // Allow double tty mode, for someone want to use virtio console
+        // for better performance, and want to keep legacy serial to catch
+        // boot stage logs for debug.
+        // Using such double tty mode, you need to configure the kernel
+        // properly, such as:
+        // "console=hvc0 earlyprintk=ttyS0"
         if self.console.mode == ConsoleOutputMode::Tty && self.serial.mode == ConsoleOutputMode::Tty
         {
-            return Err(ValidationError::DoubleTtyMode);
+            info!("Notice, {}", ValidationError::DoubleTtyMode);
```
Comment (alyssais) [inline vmm/src/config.rs:1944]:
```suggestion
            info!("{}", ValidationError::DoubleTtyMode);
```

No other info message in Cloud Hypervisor is prefixed with "Notice".

**Review by alyssais — CHANGES_REQUESTED** [review]

**Review by likebreath — COMMENTED** [review]
Can you please add an integration tests for this particular use-case?

Typo in the commit message: "enable both legacy serial `an` virito console"

File: vmm/src/config.rs (hunk 1 above)
Comment (zhuangel) [reply vmm/src/config.rs:1944]: Thanks will update.

**Comment by zhuangel** [comment]
> Can you please add an integration tests for this particular use-case?
> 
> Typo in the commit message: "enable both legacy serial `an` virito console"

Thanks for review, will update.

**Comment by rbradford** [comment]
Could we please improve the commit message "double TTY mode" doesn't really make much sense - is that a VM using a tty inside a VM also using a tty...

**Comment by zhuangel** [comment]
> Could we please improve the commit message "double TTY mode" doesn't really make much sense - is that a VM using a tty inside a VM also using a tty...

Thanks, I have updated the commit message.

**Comment by rbradford** [comment]
> > Could we please improve the commit message "double TTY mode" doesn't really make much sense - is that a VM using a tty inside a VM also using a tty...
> 
> Thanks, I have updated the commit message.

This looks a lot better thank you - but I think you meant "mode" rather than "model" in the summary.

**Comment by zhuangel** [comment]
> > > Could we please improve the commit message "double TTY mode" doesn't really make much sense - is that a VM using a tty inside a VM also using a tty...
> > 
> > 
> > Thanks, I have updated the commit message.
> 
> This looks a lot better thank you - but I think you meant "mode" rather than "model" in the summary.

Thanks, it should be "mode", updated :)

File: tests/integration.rs [hunk 2]
```diff
@@ -7093,6 +7093,58 @@ mod common_parallel {
...
     }
+
+    #[test]
+    #[cfg(target_arch = "x86_64")]
+    fn test_double_tty() {
+        let focal = UbuntuDiskConfig::new(FOCAL_IMAGE_NAME.to_string());
+        let guest = Guest::new(Box::new(focal));
+        let mut cmd = GuestCommand::new(&guest);
+        let api_socket = temp_api_path(&guest.tmp_dir);
+        let tty_str: &str = "console=hvc0 earlyprintk=ttyS0 ";
```
Comment (likebreath) [inline tests/integration.rs:7104]: @michael2012z Are you aware of anything equivalent to `earlyprintk` on `aarch64`?

File: vmm/src/config.rs [hunk 3]
```diff
@@ -1930,9 +1930,18 @@ impl VmConfig {
...
 
+        // The 'conflict' check is introduced in commit 24438e0390d3
+        // (vm-virtio: Enable the vmm support for virtio-console).
+        //
+        // Allow double tty mode, for someone want to use virtio console
+        // for better performance, and want to keep legacy serial to catch
+        // boot stage logs for debug.
+        // Using such double tty mode, you need to configure the kernel
+        // properly, such as:
+        // "console=hvc0 earlyprintk=ttyS0"
```
Comment (likebreath) [inline vmm/src/config.rs:1941]: The comments also need to be updated to avoid confusions on "double tty mode".

File: vmm/src/config.rs [hunk 4]
```diff
@@ -1930,9 +1930,18 @@ impl VmConfig {
...
         if self.console.mode == ConsoleOutputMode::Tty && self.serial.mode == ConsoleOutputMode::Tty
         {
-            return Err(ValidationError::DoubleTtyMode);
+            info!("{}", ValidationError::DoubleTtyMode);
```
Comment (likebreath) [inline vmm/src/config.rs:1944]: I think it is better to say something like: `Both serial and console are being set as TTY mode. Guest kernel command-line should include "console=hvc0 earlyprintk=ttyS0".`

Comment (zhuangel) [reply vmm/src/config.rs]: I don't think we should add linux kernel command-line suggestion in CH logs（maybe in doc or comments）, so just change the error return into some logs is enough.

File: vmm/src/config.rs (hunk 3 above)
Comment (zhuangel) [reply vmm/src/config.rs]: Thanks, updated.

File: vmm/src/config.rs [hunk 5]
```diff
@@ -1930,9 +1930,18 @@ impl VmConfig {
...
         }
```
[the remark this answers is not in the document]
Comment (zhuangel) [reply vmm/src/config.rs:1942]: Thanks, updated.

**Review by rbradford — APPROVED** [review]
I'm happy with this as is - @likebreath ?

**Review by likebreath — APPROVED** [review]

**Pull request merged**